lib/evmone/baseline.cpp

```diff
@@ -106,76 +106,94 @@ inline evmc_status_code check_requirements(
+    StopToken (*instr_fn)() noexcept, Position /*pos*/, ExecutionState& state) noexcept
 {
     state.status = instr_fn().status;
     return nullptr;
 }
 
 inline code_iterator invoke(evmc_status_code (*instr_fn)(StackTop, ExecutionState&) noexcept,
-    ExecutionState& state, code_iterator pos) noexcept
+    Position pos, ExecutionState& state) noexcept
 {
-    if (const auto status = instr_fn(state.stack.top_item, state); status != EVMC_SUCCESS)
+    if (const auto status = instr_fn(pos.stack_top, state); status != EVMC_SUCCESS)
     {
         state.status = status;
         return nullptr;
     }
-    return pos + 1;
+    return pos.code_it + 1;
 }
 
-inline code_iterator invoke(void (*instr_fn)(StackTop, ExecutionState&) noexcept,
-    ExecutionState& state, code_iterator pos) noexcept
+inline code_iterator invoke(void (*instr_fn)(StackTop, ExecutionState&) noexcept, Position pos,
+    ExecutionState& state) noexcept
 {
-    instr_fn(state.stack.top_item, state);
-    return pos + 1;
+    instr_fn(pos.stack_top, state);
+    return pos.code_it + 1;
 }
 
 inline code_iterator invoke(
-    code_iterator (*instr_fn)(StackTop, ExecutionState&, code_iterator) noexcept,
-    ExecutionState& state, code_iterator pos) noexcept
+    code_iterator (*instr_fn)(StackTop, ExecutionState&, code_iterator) noexcept, Position pos,
+    ExecutionState& state) noexcept
 {
-    return instr_fn(state.stack.top_item, state, pos);
+    return instr_fn(pos.stack_top, state, pos.code_it);
 }
 
-inline code_iterator invoke(StopToken (*instr_fn)(StackTop, ExecutionState&) noexcept,
-    ExecutionState& state, code_iterator /*pos*/) noexcept
+inline code_iterator invoke(StopToken (*instr_fn)(StackTop, ExecutionState&) noexcept, Position pos,
+    ExecutionState& state) noexcept
 {
-    state.status = instr_fn(state.stack.top_item, state).status;
+    state.status = instr_fn(pos.stack_top, state).status;
     return nullptr;
 }
 /// @}
 
 /// A helper to invoke the instruction implementation of the given opcode Op.
 template <evmc_opcode Op>
-[[gnu::always_inline]] inline code_iterator invoke(
-    const CostTable& cost_table, ExecutionState& state, code_iterator pos) noexcept
+[[gnu::always_inline]] inline Position invoke(const CostTable& cost_table,
+    const uint256* stack_bottom, Position pos, ExecutionState& state) noexcept
 {
-    if (const auto status = check_requirements<Op>(cost_table, state.gas_left, state.stack.size());
+    const auto stack_size = static_cast<int>(pos.stack_top - stack_bottom);
+    if (const auto status = check_requirements<Op>(cost_table, state.gas_left, stack_size);
         status != EVMC_SUCCESS)
     {
         state.status = status;
-        return nullptr;
+        return {nullptr, pos.stack_top};
     }
-    const auto new_pos = invoke(instr::core::impl<Op>, state, pos);
-    state.stack.top_item += instr::traits[Op].stack_height_change;
-    return new_pos;
+    return {invoke(instr::core::impl<Op>, pos, state),
```

Comment:
Nit: maybe I would leave it here with 
```
const auto new_code_it = ... 
const auti new_stack_top = ...
return {new_code_it, new_stack_top};
```
It was easier to read beforr

Reply:
Done, thanks.